src/bundle/Core/Command/CleanupVersionsCommand.php

```diff
@@ -44,18 +43,15 @@ class CleanupVersionsCommand extends Command implements BackwardCompatibleComman
         self::VERSION_PUBLISHED => VersionInfo::STATUS_PUBLISHED,
     ];
 
-    /** @var \Ibexa\Contracts\Core\Repository\Repository */
-    private $repository;
+    private Repository $repository;
 
-    /** @var \Ibexa\Bundle\Core\ApiLoader\RepositoryConfigurationProvider */
-    private $repositoryConfigurationProvider;
+    private RepositoryConfigurationProviderInterface $repositoryConfigurationProvider;
 
-    /** @var \Doctrine\DBAL\Driver\Connection */
-    private $connection;
+    private Connection $connection;
```

Comment:
```suggestion
    private readonly Connection $connection;
```

Out of curiosity, since we're touching this code - shouldn't the connection be acquired from the `Repository`? Otherwise this command would probably work on two different databases, if the `Repository` uses a different one, right?

Reply:
> Out of curiosity, since we're touching this code - shouldn't the connection be acquired from the Repository? Otherwise this command would probably work on two different databases, if the Repository uses a different one, right?

No, we want to get rid of the `Repository` entirely in the future. Injecting here repository directly creates a cyclic dependency. It was never a good pattern for a Symfony container applications.

You always need to use `@ibexa.persistence.connection` service for such connections. It's error-prone as well, but that's a separate subject. We might want to use autowiring to inject proper connection based either on an argument name (`$ibexaConnection`) or a marker interface wrapping a Doctrine `Connection` class.

Reply:
I understand that we will get rid of `Repository` soon™. The issue here is that this particular command will probably not work with multiple repositories, or will give false results - since it relies on `Connection` being swapped to the right one. Or at least I see a high likelihood of it failing.

And on the matter of `ibexa.persistence.connection`, I would point out that it also suffers from the same issue we had as with configuration for site accesses. We should get it from a similar service at runtime, since the context can change. But it's a side matter in context of this PR :sweat_smile:

Reply:
> I understand that we will get rid of `Repository` soon™. The issue here is that this particular command will probably not work with multiple repositories, or will give false results - since it relies on `Connection` being swapped to the right one. Or at least I see a high likelihood of it failing.
> 
> And on the matter of `ibexa.persistence.connection`, I would point out that it also suffers from the same issue we had as with configuration for site accesses. We should get it from a similar service at runtime, since the context can change. But it's a side matter in context of this PR 😅

1. This particular command [gets injected with `ibexa.persistence.connection`](https://github.com/ibexa/core/blob/178a0d2d9547db1e6d2656c81467d90bff24ab10/src/bundle/Core/Resources/config/services.yml#L321).
2. `ibexa.persistence.connection` is [built exactly from `StorageConnectionFactory::getConnection`](https://github.com/ibexa/core/blob/32de03cfba5dcd3e606072505dedbaf94467ab1b/src/bundle/Core/Resources/config/storage_engines.yml#L21-L25).